iptables-detect/iptables-detect.sh

```diff
@@ -0,0 +1,242 @@
+        else
+            detected_via=alternatives
+            mode=legacy
+        fi
+    fi
+}
+
+# we should not run os-detect if we're being run inside of a container
+os_detect() {
+    # perform some very rudimentary platform detection
+    lsb_dist=''
+    dist_version=''
+    if [ -z "$lsb_dist" ] && [ -r /etc/lsb-release ]; then
+        lsb_dist="$(. /etc/lsb-release && echo "$DISTRIB_ID")"
+    fi
+    if [ -z "$lsb_dist" ] && [ -r /etc/debian_version ]; then
+        lsb_dist='debian'
+    fi
+    if [ -z "$lsb_dist" ] && [ -r /etc/fedora-release ]; then
+        lsb_dist='fedora'
+    fi
+    if [ -z "$lsb_dist" ] && [ -r /etc/oracle-release ]; then
+        lsb_dist='oracleserver'
+    fi
+    if [ -z "$lsb_dist" ] && [ -r /etc/centos-release ]; then
+        lsb_dist='centos'
+    fi
+    if [ -z "$lsb_dist" ] && [ -r /etc/redhat-release ]; then
+        lsb_dist='redhat'
+    fi
+    if [ -z "$lsb_dist" ] && [ -r /etc/os-release ]; then
+        lsb_dist="$(. /etc/os-release && echo "$ID")"
+    fi
+
+    lsb_dist="$(echo "$lsb_dist" | tr '[:upper:]' '[:lower:]')"
+
+    # Special case redhatenterpriseserver
+    if [ "${lsb_dist}" = "redhatenterpriseserver" ]; then
+        # Set it to redhat, it will be changed to centos below anyways
+        lsb_dist='redhat'
+    fi
+
+    case "$lsb_dist" in
+
+    ubuntu)
+        # By default, Ubuntu is using iptables in legacy mode. Ideally, this should have been already caught by the alternatives check.
+        detected_via=os
+        mode=legacy
+        ;;
+
+    debian | raspbian)
+        dist_version="$(cat /etc/debian_version | sed 's/\/.*//' | sed 's/\..*//')"
+        # If Debian >= 10 (Buster is 10), then NFT. otherwise, assume it is legacy
+        if [ "$dist_version" -ge 10 ]; then
+            detected_via=os
+            mode=nft
+        else
+            detected_via=os
+            mode=legacy
+        fi
+        ;;
+
+    oracleserver)
+        dist_version="$(. /etc/os-release && echo "$VERSION_ID")"
+        if [ "$dist_version" -ge 8 ]; then
+            detected_via=os
+            mode=nft
+        else
+            detected_via=os
+            mode=legacy
+        fi
+        ;;
+
+    fedora)
+        # As of 05/15/2020, all Fedora packages appeared to be still `legacy` by default although there is a `iptables-nft` package that installs the nft iptables, so look for that package.
+        rpm -qa | grep -q "iptables-nft"
+        if [ $? = 0 ]; then
+            detected_via=os
+            mode=nft
+        else
+            detected_via=os
+            mode=legacy
+        fi
+        ;;
+
+    centos | redhat)
+        dist_version="$(. /etc/os-release && echo "$VERSION_ID")"
+        if [ "$dist_version" -ge 8 ]; then
+            detected_via=os
+            mode=nft
+        else
+            detected_via=os
+            mode=legacy
+        fi
+        ;;
+
+        # We are running an operating system we don't know, default to nf_tables.
```

Comment:
Not to make this more complicated than it needs to be, but is it worth it when encountering an unknown OS $ID to fallback to $ID_LIKE if it is present in `/etc/os-release`? Likely such is more appropriate as a future improvement.

Reply:
As an example, k3OS `/etc/os-release`:
```shell
NAME="k3OS"
VERSION="k3OS v0.10.2"
ID=k3os
ID_LIKE=alpine
PRETTY_NAME="k3OS v0.10.2"
VERSION_ID="v0.10.2"
HOME_URL="https://k3os.io/"
SUPPORT_URL="https://k3os.io/"
BUG_REPORT_URL="https://github.com/rancher/k3os/issues"
ISO_URL="https://github.com/rancher/k3os/releases/download/v0.10.2/k3os-amd64.iso"
```